src/coreclr/nativeaot/Runtime/amd64/WriteBarriers.S

```diff
@@ -260,8 +260,12 @@ LEAF_END RhpCheckedXchg, _TEXT
 //      rdi, rsi are incremented by 8,
 //      rcx, r10, r11: trashed
 //
+// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpByRefAssignRefAVLocation
+// - Function "UnwindSimpleHelperToCaller" assumes the stack contains just the pushed return address
 LEAF_ENTRY RhpByRefAssignRef, _TEXT
     mov     rcx, [rsi]
```

Comment:
Is it somehow guaranteed that the source is non-null and we never AV reading it?

Reply:
Looks like nothing guarantees that. Added check for that too.